[STEP4] 자동차 경주 (우승자)

SaraHan774 opened the pull request:
피드백 미리 감사드립니다!!!

SaraHan774 committed:
[step4] 기능 요구사항 리드미 정리

SaraHan774 committed:
[step4] 사용자로부터 자동차 이름을 쉼표(`,`)로 구분하여 입력받는다.

SaraHan774 committed:
[step4]  각 자동차 이름은 최대 5자까지 허용됨. 관련 테스트 추가.

SaraHan774 committed:
[step4] 매 라운드마다 각 자동차의 이름과 위치를 출력한다.

SaraHan774 committed:
[step4] 라운드가 끝난 후 최종 우승자를 출력한다.

SaraHan774 committed:
[step4] 자동차 이동 여부는 CarMovementDecider 를 통해 알도록 함

SaraHan774 committed:
[step4] RacingGameTest 테스트 수정

SaraHan774 committed:
[step4] RacingGame 은 우승자라는 지식이 없음, 최장길이 레코드만 반환하도록 수정

BeokBeok commented on src/main/kotlin/week1/racing/RacingCar.kt:
```diff
@@ -1,20 +1,25 @@
+    val name: String,
+    private val carMovementDecider: MovementDecider
+) {
+    init {
+        require(name.length <= MAX_CAR_NAME_LENGTH) {
+            "Car name is too long ! (max length = $MAX_CAR_NAME_LENGTH, current = ${name.length})"
+        }
+    }
```
init 블록을 활용한 유효성 검사 👍

BeokBeok commented on src/main/kotlin/week1/racing/RacingGame.kt:
```diff
@@ -1,30 +1,49 @@
+    val finalRecords = this.last().records
+    val maxDistance = finalRecords.maxOf { it.distance }
+    return finalRecords.filter { it.distance  == maxDistance }
+}
```
우승자를 조회하는 로직을 RacingGame 파일 내 extension으로 구현해주셨네요.
지금도 좋지만, 우승자와 관련된 책임을 갖도록 구조를 변경하여 우승자와 관련된 로직의 추가 또는 변경이 생기더라도 관련된 곳만 수정할 수 있도록 개선해보면 어떨까요?

BeokBeok commented on src/main/kotlin/week1/racing/view/InputView.kt:
```diff
@@ -1,13 +1,18 @@
 package week1.racing.view
 
 class InputView {
+    fun readStringsAndSplit(promptMessage: String, separator : String = ","): List<String> {
```
함수명만 보면 "문자열을 읽고 쪼갠다" 라고 이해가 되는데요.
함수는 한 가지 기능만 하도록 개선해보면 어떨까요?

BeokBeok commented on src/main/kotlin/week1/racing/view/ResultView.kt:
```diff
@@ -1,18 +1,32 @@
 
 class ResultView {
-
     fun printResult(rounds: List<GameRound>) {
```
함수명만 보면 "결과를 출력한다" 라고 이해가 되는데요.
출력과 관련된 요구사항이 늘어날수록 이 함수는 점점 커지게 되어 두 가지 이상의 일을 하게 될 것 같습니다. 🤔 
함수는 한 가지 일만 하도록 개선해보면 어떨까요?

BeokBeok commented on src/test/kotlin/week1/racing/RacingGameTest.kt:
```diff
@@ -5,48 +5,58 @@ import org.junit.jupiter.api.Test
+        assertEquals(nonEmptyListOfCarNames.size, gameRound.records.size)
+    }
+
+    // winner 선정 검증
+    @Test
+    fun `{given} nonEmptyListOfCarNames & nonZeroNumberOfRounds & 항상 모든 차가 전진할 경우 {when} start() {then} getMaxDistanceRecords() 사이즈 == 총 차 대수`() {
```
우승자를 간단하게 확인해볼 수 있도록 개선해보면 어떨까요?
가령, RacingCar 객체를 생성하고 이 객체들 중에 누가 우승인지 판단할 수 있다면 간단해질 것 같습니다.

BeokBeok reviewed (changes requested):
이전 단계 피드백과 자동차 경주(우승자) 미션하시느라 고생하셨습니다. 👍 
고민해볼만한 의견들을 코맨트로 작성하였으니, 충분히 고민해보시고 도전해보세요. 💪

SaraHan774 committed:
[step4] 코드리뷰 반영
- InputView 의 함수 역할 쪼개기
- ResultView 의 함수 역할 쪼개기
- RacingGame 이 우승자 명단을 반환하도록 API 수정
- RacingGameTest 수정

SaraHan774 committed:
Merge

BeokBeok commented on src/main/kotlin/week1/racing/RacingGame.kt:
```diff
@@ -29,14 +29,17 @@ class RacingGame(
+        val winnerRecords = records.filter { it.distance == maxDistance }
+        return winnerRecords
+    }
 }
```
가희님이 생각하시는 RacingGame의 책임은 어디까지라고 생각하시나요? 🤔 
우승자를 관리하는 클래스를 만들어보면 어떨까요?

SaraHan774 replied:
저도 우승자를 RacingGame 이 알고있는게 어색하게 느껴졌었는데요, 
피드백 반영해 보도록 하겠습니다!

BeokBeok reviewed (approved):
피드백 반영하시느라 고생하셨습니다. 👍 
추가로 고민해볼만한 의견들을 코맨트로 작성하였으니, 충분히 고민해보시고 도전해보세요. 💪 
다음 미션을 진행하시면서, 추가로 작성된 코맨트들도 반영해주세요. 🙏
